frappe/public/js/frappe/form/toolbar.js

```diff
@@ -103,10 +102,29 @@ frappe.ui.form.Toolbar = class Toolbar {
 				docname,
 				name: new_name,
 				title: new_title,
+				enqueue: true,
 				merge,
 				freeze: true,
 				freeze_message: __("Updating related fields...")
 			}).then(new_docname => {
+				// handle document renaming queued action
+				if (new_docname == docname) {
+					frappe.socketio.doc_subscribe(doctype, new_name);
+					frappe.realtime.on("doc_update", data => {
+						if (data.doctype == doctype && data.name == new_name) {
+							$(document).trigger("rename", [doctype, docname, new_name]);
+							if (locals[doctype] && locals[doctype][docname]) delete locals[doctype][docname];
+							this.frm.reload_doc();
+							frappe.show_alert({
+								message: __('Document renamed from {0} to {1}', [docname.bold(), new_name.bold()]),
+								indicator: 'success',
+							});
+						}
+					});
+					frappe.show_alert(
+						__('Document renaming from {0} to {1} has been queued', [docname.bold(), new_name.bold()])
```

Comment:
```suggestion
						__('Document renaming from {0} to {1} has been queued, please check after some time.', [docname.bold(), new_name.bold()])
```

"queued" is bit technical, can we add a bit more info?

---

Minor UX issues around locking:

Immediately re-renaming is allowed even if the previous renaming isn't finished. (poor locking implementation I guess?) Perhaps we should immediately make the form read-only or `frm.disable_save` on queued docs with some message for why it's read-only. I think we need to utilize `HTTP 202` for this 😄 


PS:  I just added 100s sleep to simulate long renames.

Reply:
PS: ERPNext has some longish complicated rename tests if reqd.

Reply:
Another thing to consider. Some renames have [pre-rename validations](https://github.com/ankush/erpnext/blob/90e4b0bdd9746590bd44a3d622168a2f98bb5129/erpnext/stock/doctype/item/item.py#L395-L402), right now it seems the messages are just posted as comments on doc.

The problem is we can't just run before_rename too since that can also have some non-validation actions => partial commits 😐

Reply:
> "queued" is bit technical, can we add a bit more info?

We could; but I'm not sure about this. When the renaming is completed successfully, the user is re-directed to the document. I'm not sure about that either since we've not really standardized these things yet 🤔. Hoping that enqueuing actions becomes first-class soon in Frappe 🥲 

> Immediately re-renaming is allowed even if the previous renaming isn't finished. (poor locking implementation I guess?) Perhaps we should immediately make the form read-only or `frm.disable_save` on queued docs with some message for why it's read-only. I think we need to utilize `HTTP 202` for this 😄

Couldn't replicate this. Did you restart your bench before trying this?

> Another thing to consider. Some renames have [pre-rename validations](https://github.com/ankush/erpnext/blob/90e4b0bdd9746590bd44a3d622168a2f98bb5129/erpnext/stock/doctype/item/item.py#L395-L402), right now it seems the messages are just posted as comments on doc.
> 
> The problem is we can't just run before_rename too since that can also have some non-validation actions => partial commits 😐

Moved validations out of the queued job, made sure they just run once whether in or out. Thought of throwing an exception if a commit was triggered in the `before_rename` but I feel I may have already gone beyond the scope of this PR.